### PR TITLE
Implement BoundPredicateVisitor trait for ManifestFilterVisitor

### PR DESCRIPTION
**Pull request by s-akhtar-baig**
GitHub issue: https://github.com/apache/iceberg-rust/issues/350

Description: ManifestEvaluator was defined in https://github.com/apache/iceberg-rust/pull/322. This PR implements all functions and adds most of the Python unit tests. 

Testing: Added new unit tests.

### CODE REVIEW

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 1]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+                            "Cannot perform not_starts_with on non-string upper bound",
+                        ));
+                    };
+
+                    // if upper is shorter than the prefix then upper can't start with the prefix
+                    if prefix_len > upper_bound.chars().count() {
+                        return ROWS_MIGHT_MATCH;
+                    }
+
+                    let truncated_upper_bound = String::from(&upper_bound[..prefix_len]);
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:338]: You can't safely use `prefix_len` to index into `upper_bound` as there is no guarantee that the index will fall on a valid UTF8 character boundary, and if it doesn't then Rust will panic.

See here for an example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e4950cb298bd87560be74fa48576184

You need to do this instead:

```rust
let truncated_upper_bound = upper_bound.chars().take(prefix_len).collect::<String>();
```

(See here for working example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=86969f315ef1bdc4a0863bde9356c1f4)

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 2]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
     ) -> crate::Result<bool> {
-        Ok(self
-            .field_summary_for_reference(reference)
-            .contains_nan
-            .is_some())
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if let Some(contains_nan) = field.contains_nan {
+            Ok(contains_nan)
+        } else {
+            ROWS_MIGHT_MATCH
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:134]: The Java implementation also tests if all values are null at this point - see here: https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java#L151

You could extract [some of the code in not_null](https://github.com/apache/iceberg-rust/pull/367/files?diff=unified&w=0#diff-f3185efcdea2022bdedc017f7859727f2e44ac0c004f08ad6a96fdfa5fa6f69aR112-R120) into a shared method, `contains_nulls_only`, and call that here.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 3]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if let Some(upper_bound) = &field.upper_bound {
+            Ok(upper_bound >= &Literal::from(datum.clone()))
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:205]: You can avoid the clone here, and in the other inequality methods:

```rust
if let Some(Literal::Primitive(upper_bound)) = &field.upper_bound {
    Ok(upper_bound >= datum.literal())
} else {
    ROWS_CANNOT_MATCH
}
```

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 4]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+        if let Some(lower_bound) = &field.lower_bound {
+            let Literal::Primitive(PrimitiveLiteral::String(lower_bound)) = lower_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string lower bound",
+                ));
+            };
+
+            let min_len = lower_bound.chars().count().min(prefix_len);
+            let truncated_lower_bound = String::from(&lower_bound[..min_len]);
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:262]: as with earlier comment, indexing into strings will not work in many situations.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 5]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+        if let Some(upper_bound) = &field.upper_bound {
+            let Literal::Primitive(PrimitiveLiteral::String(upper_bound)) = upper_bound else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string upper bound",
+                ));
+            };
+
+            let min_len = upper_bound.chars().count().min(prefix_len);
+            let truncated_upper_bound = String::from(&upper_bound[..min_len]);
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:277]: String indexing

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 6]
```diff
@@ -98,103 +101,249 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+                    "Cannot perform not_starts_with on non-string lower bound",
+                ));
+            };
+
+            // if lower is shorter than the prefix then lower doesn't start with the prefix
+            if prefix_len > lower_bound.chars().count() {
+                return ROWS_MIGHT_MATCH;
+            }
+
+            let truncated_lower_bound = String::from(&lower_bound[..prefix_len]);
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:322]: String indexing

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 7]
```diff
@@ -203,16 +352,42 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+            return ROWS_CANNOT_MATCH;
+        }
+
+        if literals.len() > IN_PREDICATE_LIMIT {
+            return ROWS_MIGHT_MATCH;
+        }
+
+        if literals
+            .iter()
+            .all(|datum| field.lower_bound.as_ref().unwrap() > &Literal::from(datum.clone()))
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:366]: We should not have any `unwrap`s in here. We don't want to panic if we encounter an unexpected situation, we should return an `Err` instead.

Also, you're cloning `datum` and creating a `Literal` every time you go through the loop here.

I'd suggest something more like this:

```rust
if let Some(lower_bound) = &field_summary.lower_bound {
    let Literal::Primitive(PrimitiveLiteral::String(lower_bound)) = lower_bound else {
        return Err(Error::new(
            ErrorKind::Unexpected,
            "Cannot use StartsWith operator on non-string lower_bound value",
        ));
    };

    let lower_bound_lit = PrimitiveLiteral::String(lower_bound.to_string());

    let all_literals_less_than_lower_bound = literals
        .iter()
        .all(|lit| lit.literal().lt(&lower_bound_lit));

    if all_literals_less_than_lower_bound {
        return ROWS_CANT_MATCH;
    }
}
```

**Review by sdd — CHANGES_REQUESTED** [review]
Looks great, thanks so much for the contribution! Just a few small issues that are straightforward to resolve. 🙌🏼

**Comment by s-akhtar-baig** [comment]
@sdd, thank you for reviewing the changes and providing references! I have modified my code based on your suggestions. Please take a look and let me know if I miss anything.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 8]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
         &mut self,
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if let Some(Literal::Primitive(lower_bound)) = &field.lower_bound {
+            Ok(lower_bound < datum.literal())
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:163]: It's a bit inconsistent to return `ROWS_MIGHT_MATCH` / `ROWS_CANNOT_MATCH` in most places, but then in some places to return in this style. To improve readability and consistency, could you refactor this, the other inequality methods, and `eq` to return `ROWS_MIGHT_MATCH` / `ROWS_CANNOT_MATCH` rather than `Ok(...)` please? I know that adds a few extra lines but it improves readability.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 9]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
         &mut self,
         reference: &BoundReference,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        Ok(self
-            .field_summary_for_reference(reference)
-            .contains_nan
-            .is_some())
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if field.contains_nan.is_some_and(|x| x == false) {
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:127]: Generally this project's style is to avoid having to disable any clippy lints.

You could achieve this here (and similarly in `not_nan` by refactoring this slightly:

```rust
if let Some(contains_nan) = field.contains_nan {
    if !contains_nan {
        return ROWS_CANNOT_MATCH;
    }
}
```

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 10]
```diff
@@ -221,67 +394,213 @@ impl ManifestFilterVisitor<'_> {
         let pos = reference.accessor().position();
         &self.partitions[pos]
     }
+
+    #[allow(clippy::bool_comparison)]
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:398]: Similar to prev comment, please refactor to avoid needing to disable this clippy lint.

**Review by sdd — CHANGES_REQUESTED** [review]
We're almost there! Just a couple of small stylistic changes required and then I'm happy. 

Thanks again! 😁

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 11]
```diff
@@ -203,16 +353,39 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if field.lower_bound.is_none() {
+            return ROWS_CANNOT_MATCH;
+        }
+
+        if literals.len() > IN_PREDICATE_LIMIT {
+            return ROWS_MIGHT_MATCH;
+        }
+
+        if let Some(Literal::Primitive(lower_bound)) = &field.lower_bound {
```
Comment (sdd) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:365]: This is much cleaner! Thanks :-)

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 12]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+        let PrimitiveLiteral::String(prefix) = datum.literal() else {
+            return Err(Error::new(
+                ErrorKind::Unexpected,
+                "Cannot perform starts_with on non-string value",
+            ));
+        };
+
+        let prefix_len = prefix.chars().count();
+
+        if let Some(lower_bound) = &field.lower_bound {
```
Comment (marvinlanhenke) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:253]: suggestion: extract into helper fn, since its used in multiple places?

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 13]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+                            "Cannot perform not_starts_with on non-string upper bound",
+                        ));
+                    };
+
+                    // if upper is shorter than the prefix then upper can't start with the prefix
+                    if prefix_len > upper_bound.chars().count() {
+                        return ROWS_MIGHT_MATCH;
+                    }
+
+                    let truncated_upper_bound =
```
Comment (marvinlanhenke) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:338]: I think we can avoid the extra String allocation by using `char_indices()` to get the prefix_len and then use a slice for comparison `let truncated_upper_bound = &upper_bound[..prefix_len];` // haven't tested it though

**Review by marvinlanhenke — CHANGES_REQUESTED** [review]
Thanks for this PR @s-akhtar-baig which looks really good - left some minor comments. However, please check the comment about 'comparison' and the `and` implementation and verify its correct.

**Review by s-akhtar-baig — COMMENTED** [review]
@marvinlanhenke @sdd, thank you for reviewing. I have modified the code accordingly and added comments for clarification. Let me know what you think.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 14]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
         reference: &BoundReference,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+
+        // contains_null encodes whether at least one partition value is null,
+        // lowerBound is null if all partition values are null
+        if self.are_all_null(field, &reference.field().field_type) {
```
[the remark this answers is not in the document]
Comment (marvinlanhenke) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:113]: ups sorry my mistake here

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs (hunk 13 above)
Comment (sdd) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:357]: I think @marvinlankenke's suggestion is indeed a better one than mine here

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 15]
```diff
@@ -103,98 +106,245 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
-        Ok(self
-            .field_summary_for_reference(reference)
-            .contains_nan
-            .is_some())
+        let field: &FieldSummary = self.field_summary_for_reference(reference);
+        if field.contains_nan.is_some_and(|x| x == false) {
+            return ROWS_CANNOT_MATCH;
+        }
+
+        if self.are_all_null(field, &reference.field().field_type) {
```
[the remark this answers is not in the document]
Comment (sdd) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:132]: @marvinlanhenke if field.contains_nan is `None` rather than `Some(false)` then this implies that the metrics don't indicate if the fields contain NaN or not. So the subsequent check for all nulls is still valid?

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 16]
```diff
@@ -103,98 +106,264 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field = self.field_summary_for_reference(reference);
+
+        if field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_CANNOT_MATCH;
```
[the remark this answers is not in the document]
Comment (s-akhtar-baig) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:222]: @liurenjie1024, I followed Python implementation https://github.com/apache/iceberg-python/blob/20f6afdf5f000ea5b167e804012f2000aa5b8573/pyiceberg/expressions/visitors.py#L639. 

Please let me know if this is incorrect and there is a different spec that I needed to follow.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 17]
```diff
@@ -103,98 +106,264 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
     fn not_eq(
         &mut self,
-        reference: &BoundReference,
-        literal: &Datum,
+        _reference: &BoundReference,
+        _datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        // because the bounds are not necessarily a min or max value, this cannot be answered using
```
[the remark this answers is not in the document]
Comment (s-akhtar-baig) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:246]: Same comment here, followed https://github.com/apache/iceberg-python/blob/20f6afdf5f000ea5b167e804012f2000aa5b8573/pyiceberg/expressions/visitors.py#L658.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 18]
```diff
@@ -103,98 +106,264 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
         reference: &BoundReference,
-        literal: &Datum,
+        datum: &Datum,
         _predicate: &BoundPredicate,
     ) -> crate::Result<bool> {
-        todo!()
+        let field = self.field_summary_for_reference(reference);
+
+        if field.lower_bound.is_none() || field.upper_bound.is_none() {
+            return ROWS_CANNOT_MATCH;
```
[the remark this answers is not in the document]
Comment (s-akhtar-baig) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:260]: Same comment here. Followed https://github.com/apache/iceberg-python/blob/20f6afdf5f000ea5b167e804012f2000aa5b8573/pyiceberg/expressions/visitors.py#L731. Collapsed if statements on L722 and L731.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs (hunk 13 above)
Comment (sdd) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:357]: On reflection, I think we can just do this:

```rust
if prefix.as_bytes().eq(&lower_bound.as_bytes()[..prefix.len()]) {
```

Since we know `prefix` is correctly formed UTF-8 as it is a `&String`, if all we are doing is comparing equality of the first n chars, if all of the initial bytes of the bound are the same as in `prefix` then the slice must also fall on a char boundary.

**Comment by s-akhtar-baig** [comment]
@liurenjie1024 @marvinlanhenke @sdd, thank you for reviewing these changes and for your feedback! Greatly appreciated!

I have pushed the required changes. Please let me know if I missed anything, thanks!

**Review by sdd — APPROVED** [review]
This looks good to me! Thanks for persevering :-)

**Comment by sdd** [comment]
Hi @liurenjie1024 - sorry to pester you but are you able to re-review this please? It's the last major piece of the puzzle on the  read side.

**Review by marvinlanhenke — APPROVED** [review]
@s-akhtar-baig 
LGTM. Thanks for your effort here, just a minor nit that is still unresolved from the last review about extracting a common helper function, but this is not a blocker.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs [hunk 19]
```diff
@@ -103,98 +106,255 @@ impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
...
+            }
+        }
+
+        if let Some(upper_bound) = &field.upper_bound {
+            let PrimitiveLiteral::String(upper_bound) = upper_bound.literal() else {
+                return Err(Error::new(
+                    ErrorKind::Unexpected,
+                    "Cannot perform starts_with on non-string upper bound",
+                ));
+            };
```
Comment (s-akhtar-baig) [inline crates/iceberg/src/expr/visitors/manifest_evaluator.rs:289]: @marvinlanhenke, I looked at reusing some of the code, but because we have custom checks, I could only extract this portion of the code (e.g. lines 283 to 289) for lower bound and upper bound in separate utility functions. Once I do that then, I still have to unwrap the returned value if it wasn't an error. The new code doesn't make the existing code much cleaner, in my opinion. But if you think it would add value then, I can push the new changes.

Comment (marvinlanhenke) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:289]: > (e.g. lines 283 to 289)

@s-akhtar-baig those were the lines I was referring to. 

Something like that might work?

```Rust
// mostly pseudo code
fn check_boundaries(field: &FieldSummary) -> Result<&String> {
  match &field.lower_bound {
    Some(lower_bound) => {
      let PrimitiveLiteral::String(lower_bound) = lower_bound.literal() else {
        return Err(Error::new(ErrorKind::Unexpected, "Cannot perform..."));
      }
      Ok(lower_bound)
    }
   None => Err(Error::new(...))
  }
}
```

If that works you can probably create a small macro to avoid having a lower and upper bound variant?
```Rust
macro_rules! check_boundaries {
    ($field:expr, $boundary:ident) => {
        match &$field.$boundary {
            Some(bound_value) => {
//...the rest
```

Comment (marvinlanhenke) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:289]: > (e.g. lines 283 to 289)

@s-akhtar-baig those were the lines I was referring to. 
Perhaps something like this might work?

```Rust
// this should be enough?
fn check_bounds(datum: &Datum ) -> Result<&String> {
  let PrimitiveLiteral::String(boundary) = datum.literal() else {
    return Err(Error::new(ErrorKind::Unexpected, "Cannot perform..."));
  };
  Ok(boundary)
}

// and then
if let Some(upper_bound) = &field.upper_bound {
  let boundary = Self::check_bounds(upper_bound)?;
  // ...proceed with custom checks
}
```

Comment (s-akhtar-baig) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs]: Thanks @marvinlanhenke! Please let me know if the latest changes reflect your suggestions correctly.

Comment (marvinlanhenke) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs]: > Thanks @marvinlanhenke! Please let me know if the latest changes reflect your suggestions correctly.

Yes this looks great 👌 thanks for your effort.

File: crates/iceberg/src/expr/visitors/manifest_evaluator.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Fokko) [reply crates/iceberg/src/expr/visitors/manifest_evaluator.rs:243]: There is a fair chance they are truncated, so you still cannot say if they are not equal.

**Review by Fokko — APPROVED** [review]
Thanks @s-akhtar-baig for working on this, and adding all the tests 🙌  Thanks @marvinlanhenke, @sdd and @liurenjie1024 for the review!
